File: HW/HW9/TreeNode.java
import java.util.Scanner;

public class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) { this.val = val; }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    public static TreeNode createTree(){
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(3);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(1);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);
        return root;
    }

    public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        if(root == null) 
            return null;

        int pValue = p.val;
        int qValue = q.val;

        TreeNode current = root;
        while(current != null) {
            if (pValue < current.val && qValue < current.val){
                current = current.left;
            }
            else if (pValue > current.val && qValue > current.val){
                current = current.right;
            }
            else{
                return current;
            }
        }
        return null;
    }

    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);

        TreeNode root = createTree();

        System.out.print("Enter the value of p: ");
        int pValue = scanner.nextInt();
        TreeNode p = new TreeNode(pValue);

        System.out.print("Enter the value of q: ");
        int qValue = scanner.nextInt();
        TreeNode q = new TreeNode(qValue);

        TreeNode treeNode = new TreeNode();  // Create an instance of TreeNode
        TreeNode lca = treeNode.lowestCommonAncestor(root, p, q);  // Call the method using the instance
        System.out.println("Lowest Common Ancestor: " + (lca != null ? lca.val : "null"));
    }
}

File: HW/HW9/TreeNode.py
class TreeNode(object):

    def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
         
    def create_tree(self):
        root = TreeNode(4)
        root.left = TreeNode(3)
        root.right = TreeNode(8)
        root.left.left = TreeNode(1)
        root.right.left = TreeNode(5)
        root.right.right = TreeNode(9)
        return root

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

        if root is None:
            return None
        
        p_value = p.val
        q_value = q.val
        
        current = root
        while current:
            if p_value < current.val and q_value < current.val:
                current = current.left
            elif p_value > current.val and q_value > current.val:
                current = current.right
            else: 
                return current
        return None
    
def main():
    tree_node = TreeNode()
    root = tree_node.create_tree()
    
    p_value = int(input("Enter the value of p: "))
    p = TreeNode(p_value)
    
    q_value = int(input("Enter the value of q: "))
    q = TreeNode(q_value)
    
    lca = tree_node.lowestCommonAncestor(root, p, q)
    print("Lowest Common Ancestor:", lca.val if lca else "null")
    
if __name__ == "__main__":
    main()   
